File: work/compaction-floors/floors.ts
// Cheap keep-signal floors for compaction (bead jev-al06), on the two labelled development sets
// (jev-x86y and jev-jec6). Definitions are preregistered in
// docs/demos/upstream-repro/compaction-floors-20260924.md. No Jev call. Writes numbers and ids only.
//
//   node --experimental-strip-types work/compaction-floors/floors.ts   # writes features-x86y.jsonl, features-jec6.jsonl
//
// Per prefix call, from the same cut the two readouts used (need.ts cut()):
//   position     the call's 1-based order among the prefix's paired calls (higher = more recent)
//   tool_class   2 read/grep/glob, 1 bash/eval, 0 anything else (write, edit, todo, hub, task, wait)
//   result_chars the result's length in characters
//   later_overlap  distinct identifiers from the call's input that appear in any later PREFIX message
//                  (visible at compaction time)
//   horizon_overlap  the same identifiers appearing in any HORIZON message. A reference only: it reads
//                  the future, so no compactor can use it.
import { createHash, timingSafeEqual } from "node:crypto";
import { readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import type { Message } from "../../fast-jev-compaction/dist/index.js";
import { cut, load, untilde } from "../compaction-need/need.ts";

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = join(HERE, "..", "..");
const SETS = [
  { name: "x86y", dir: join(ROOT, "work/compaction-need"), excluded: join(ROOT, "work/compaction-need/excluded.json") },
  { name: "jec6", dir: join(ROOT, "work/compaction-keep"), excluded: null },
];
const READS = new Set(["read", "grep", "glob"]);
const RUNS = new Set(["bash", "eval"]);
// An identifier: a run of 6+ path/name characters holding at least one of / . _ (paths, file names, ids).
const IDENT = /[A-Za-z0-9_./-]{6,}/g;

function identifiers(input: unknown): Set<string> {
  const out = new Set<string>();
  for (const m of JSON.stringify(input).matchAll(IDENT)) if (/[/._]/.test(m[0])) out.add(m[0]);
  return out;
}

function messageText(m: Message): string {
  return [m.text ?? "", ...(m.toolUses ?? []).map((u) => JSON.stringify(u.input)), ...(m.toolResults ?? []).map((r) => r.text ?? "")].join("\n");
}

function hits(ids: Set<string>, texts: readonly string[]): number {
  let n = 0;
  for (const id of ids) if (texts.some((t) => t.includes(id))) n += 1;
  return n;
}

for (const set of SETS) {
  const sessions = JSON.parse(readFileSync(join(set.dir, "sessions.json"), "utf8")).sessions;
  const excluded: Record<string, string> = set.excluded ? JSON.parse(readFileSync(set.excluded, "utf8")) : {};
  const rows = [];
  for (const s of sessions) {
    if (s.id in excluded) continue; // never read (jev-x86y amendment A1)
    const bytes = readFileSync(untilde(s.path));
    if (!timingSafeEqual(createHash("sha256").update(bytes).digest(), Buffer.from(s.sha256, "hex"))) {
      throw new Error(`REFUSED: ${s.path} changed since it was sampled`);
    }
    const c = cut(load(untilde(s.path)))!;
    const horizonTexts = c.horizon.map(messageText);
    const results = new Map<string, number>();
    for (const m of c.prefix) for (const r of m.toolResults ?? []) results.set(r.tool_use_id, (r.text ?? "").length);
    c.prefixCalls.forEach((call, k) => {
      const ids = identifiers(call.input);
      const later = c.prefix.slice(call.resultIndex + 1).map(messageText);
      rows.push({
        session: s.id,
        tool_use_id: call.tool_use_id,
        pinned: call.pinned,
        position: k + 1,
        tool_class: READS.has(call.tool) ? 2 : RUNS.has(call.tool) ? 1 : 0,
        result_chars: results.get(call.tool_use_id) ?? 0,
        later_overlap: hits(ids, later),
        horizon_overlap: hits(ids, horizonTexts),
      });
    });
  }
  writeFileSync(join(HERE, `features-${set.name}.jsonl`), rows.map((r) => JSON.stringify(r)).join("\n") + "\n");
  console.log(`${set.name}: ${rows.length} prefix calls`);
}

File: work/compaction-floors/floors.py
#!/usr/bin/env python3
"""Cheap keep-signal floors vs Jev's keep scores (bead jev-al06). Development sets only, keyless.

Definitions are preregistered in docs/demos/upstream-repro/compaction-floors-20260924.md.
features-<set>.jsonl comes from floors.ts; labels and Jev scores come from each readout's own files.

  python3 work/compaction-floors/floors.py
"""

import json
import os
import random
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.join(HERE, "..", "..")
sys.path.insert(0, os.path.join(ROOT, "work", "compaction-need"))
sys.path.insert(0, os.path.join(ROOT, "work", "compaction-keep"))
import need as N  # noqa: E402
import keep as K  # noqa: E402

FLOORS = ["position", "tool_class", "result_chars", "later_overlap"]
REFERENCE = ["horizon_overlap"]  # reads the future: no compactor can use it
DROP_RATE = 0.40  # C1's measured not-needed drop rate on jec6
SEED = 20260924
RESAMPLES = 2000


def auc(pos, neg):
    """Mann-Whitney AUC by midranks: P(a needed score > a not-needed score), ties counted half."""
    if not pos or not neg:
        return None
    ranked = sorted([(v, 1) for v in pos] + [(v, 0) for v in neg])
    rank_sum, i = 0.0, 0
    while i < len(ranked):
        j = i
        while j < len(ranked) and ranked[j][0] == ranked[i][0]:
            j += 1
        mid = (i + 1 + j) / 2
        rank_sum += mid * sum(flag for _v, flag in ranked[i:j])
        i = j
    return (rank_sum - len(pos) * (len(pos) + 1) / 2) / (len(pos) * len(neg))


def auc_ci(pos, neg):
    """Percentile bootstrap 95% interval, resampling each class, seeded."""
    rng = random.Random(SEED)
    vals = sorted(
        auc([rng.choice(pos) for _ in pos], [rng.choice(neg) for _ in neg])
        for _ in range(RESAMPLES)
    )
    return vals[int(0.025 * RESAMPLES)], vals[int(0.975 * RESAMPLES) - 1]


def at_drop_rate(pos, neg):
    """The cut that drops ~40% of not-needed calls (the 40th percentile of their scores); needed kept at it."""
    ordered = sorted(neg)
    cut = ordered[int(DROP_RATE * len(ordered))]
    return sum(p >= cut for p in pos), sum(n < cut for n in neg) / len(neg)


def load_set(name):
    if name == "x86y":
        keys = N.calls()
        got = N.final(keys)
        dec = {(d["session"], d["tool_use_id"]): d for d in N.read_jsonl(N.DECISIONS)}
    else:
        keys = K.calls()
        got = K.final(keys)
        dec = {(d["session"], d["tool_use_id"]): d for d in N.read_jsonl(K.DECISIONS)}
    if isinstance(got, str):
        raise SystemExit(f"REFUSED: {name} labels not final: {got}")
    final = got[2]
    feats = {
        (r["session"], r["tool_use_id"]): r
        for r in N.read_jsonl(os.path.join(HERE, f"features-{name}.jsonl"))
    }
    rows = []
    for k in keys:
        if keys[k]["pinned"] or final[k] == "undecidable":
            continue
        if k not in feats or feats[k]["pinned"] != keys[k]["pinned"]:
            raise SystemExit(f"REFUSED: {name} features do not match calls.json at {k}")
        row = dict(feats[k])
        row["needed"] = final[k] == "needed"
        row["keepResult"] = dec[k]["keepResult"]
        row["keepCall"] = dec[k]["keepCall"]
        if "need" in dec[k]:
            row["need"] = dec[k]["need"]
        rows.append(row)
    return rows


def main():
    print(
        "DEVELOPMENT-SET COMPARISON: every row below was labelled and scored before; nothing here is held out.\n"
    )
    for name in ("x86y", "jec6"):
        rows = load_set(name)
        pos = [r for r in rows if r["needed"]]
        neg = [r for r in rows if not r["needed"]]
        print(
            f"## {name}: {len(pos)} needed, {len(neg)} not-needed unpinned calls (undecidable and pinned left out)"
        )
        print(
            "| signal | kind | AUC | bootstrap 95% | needed kept at a 40% not-needed drop | actual drop |"
        )
        print("|---|---|---:|---|---:|---:|")
        jev = ["keepResult", "keepCall"] + (["need"] if "need" in rows[0] else [])
        for sig, kind in (
            [(s, "Jev") for s in jev]
            + [(s, "floor") for s in FLOORS]
            + [(s, "reference (uses the horizon)") for s in REFERENCE]
        ):
            p = [r[sig] for r in pos]
            n = [r[sig] for r in neg]
            a = auc(p, n)
            lo, hi = auc_ci(p, n)
            kept, drop = at_drop_rate(p, n)
            print(
                f"| {sig} | {kind} | {a:.3f} | {lo:.3f}-{hi:.3f} | {kept}/{len(p)} | {drop:.0%} |"
            )
        print()
    return 0


if __name__ == "__main__":
    sys.exit(main())
